examples/blake2f.rs

```diff
@@ -0,0 +1,1371 @@
+        .map(|i| ctx.forward(format!("v_vec[{}]", i).as_str()))
+        .collect();
+    let h_vec: Vec<Queriable<F>> = (0..H_LEN)
+        .map(|i| ctx.forward(format!("h_vec[{}]", i).as_str()))
+        .collect();
+    let m_vec: Vec<Queriable<F>> = (0..M_LEN)
+        .map(|i| ctx.forward(format!("m_vec[{}]", i).as_str()))
+        .collect();
+    let round = ctx.forward("round");
+
+    let blake2f_pre_step = ctx.step_type_def("blake2f_pre_step", |ctx| {
+        let v_vec = v_vec.clone();
+        let setup_v_vec = v_vec.clone();
+
+        let h_vec = h_vec.clone();
+        let setup_h_vec = h_vec.clone();
+
+        let m_vec = m_vec.clone();
+        let setup_m_vec = m_vec.clone();
+
+        let t0 = ctx.internal("t0");
+        let t1 = ctx.internal("t1");
+        let f = ctx.internal("f");
+
+        let h_split_4bits_vec: Vec<Vec<Queriable<F>>> = (0..H_LEN)
+            .map(|i| {
+                (0..BITS_NUMBER)
+                    .map(|j| ctx.internal(format!("h_split_4bits_vec[{}][{}]", i, j).as_str()))
+                    .collect()
+            })
+            .collect();
+        let setup_h_split_4bits_vec = h_split_4bits_vec.clone();
+
+        let m_split_4bits_vec: Vec<Vec<Queriable<F>>> = (0..M_LEN)
+            .map(|i| {
+                (0..BITS_NUMBER)
+                    .map(|j| ctx.internal(format!("m_split_4bits_vec[{}][{}]", i, j).as_str()))
+                    .collect()
+            })
+            .collect();
+        let setup_m_split_4bits_vec = m_split_4bits_vec.clone();
+
+        let t_split_4bits_vec: Vec<Vec<Queriable<F>>> = (0..2)
+            .map(|i| {
+                (0..BITS_NUMBER)
+                    .map(|j| ctx.internal(format!("t_split_4bits_vec[{}][{}]", i, j).as_str()))
+                    .collect()
+            })
+            .collect();
+        let setup_t_split_4bits_vec = t_split_4bits_vec.clone();
+
+        let iv_split_4bits_vec: Vec<Vec<Queriable<F>>> = (0..3)
+            .map(|i| {
+                (0..BITS_NUMBER)
+                    .map(|j| ctx.internal(format!("iv_split_4bits_vec[{}][{}]", i, j).as_str()))
+                    .collect()
+            })
+            .collect();
+        let setup_iv_split_4bits_vec = iv_split_4bits_vec.clone();
+
+        let final_split_bits_vec: Vec<Vec<Queriable<F>>> = (0..3)
+            .map(|i| {
+                (0..BITS_NUMBER)
+                    .map(|j| ctx.internal(format!("final_split_bits_vec[{}][{}]", i, j).as_str()))
+                    .collect()
+            })
+            .collect();
+        let setup_final_split_bits_vec = final_split_bits_vec.clone();
+
+        ctx.setup(move |ctx| {
+            // check inputs: h_vec
+            for (i, split_4bits_vec) in h_split_4bits_vec.iter().enumerate() {
+                let mut h_4bits_sum_value = 0.expr() * 1;
+                for &bits in split_4bits_vec.iter().rev() {
+                    ctx.add_lookup(params.bits_table.apply(bits).apply(bits));
```

Comment:
You could make a function that is called `check_4bit(ctx, a)` and does `ctx.add_lookup(params.bits_table.apply(bits).apply(bits));` that would be more readable.

Reply:
Ok, Got it.